### PR TITLE
lease: do lease pile-up reduction in the background

### PR DESCRIPTION
**Pull request by mgates**
This moves lease pile reduction into a goroutine. This prevents timeouts
when the lessor is locked for a long time (when there are a lot of
leases, mostly).

We had a problem where when we had a lot of leases (100kish), and a leader election happened the lessor was locked for a long time causing timeouts when we tried to grant a lease. This changes it so that the lessor is locked in batches, which allows for creation of leases while the leader is initializing the lessor. It still won't start expiring leases until it's all done rewriting them, though.

Before:
```
BenchmarkLessorPromote1-16                        500000 4036 ns/op
BenchmarkLessorPromote10-16                       500000 3932 ns/op
BenchmarkLessorPromote100-16                      500000 3954 ns/op
BenchmarkLessorPromote1000-16                     300000 3906 ns/op
BenchmarkLessorPromote10000-16                    300000 4639 ns/op
BenchmarkLessorPromote100000-16                      100 27216481 ns/op
BenchmarkLessorPromote1000000-16                     100 325164684 ns/op
```

After:
```
BenchmarkLessorPromote1-16                        500000 3769 ns/op
BenchmarkLessorPromote10-16                       500000 3835 ns/op
BenchmarkLessorPromote100-16                      500000 3829 ns/op
BenchmarkLessorPromote1000-16                     500000 3665 ns/op
BenchmarkLessorPromote10000-16                    500000 3800 ns/op
BenchmarkLessorPromote100000-16                   300000 4114 ns/op
BenchmarkLessorPromote1000000-16                  300000 5143 ns/op
```

### CODE REVIEW

**Comment by mgates** [comment]
I'll have some more realistic test examples on Monday.

**Comment by cosgroveb** [comment]
This addresses #9496

File: lease/lessor.go [hunk 1]
```diff
@@ -329,66 +332,82 @@ func (le *lessor) unsafeLeases() []*Lease {
...
 	le.mu.RUnlock()
+
+	sort.Sort(leasesByExpiry(ls))
 	return ls
 }
 
 func (le *lessor) Promote(extend time.Duration) {
-	le.mu.Lock()
-	defer le.mu.Unlock()
+	le.Ready = false
```
Comment (gyuho) [inline lease/lessor.go]: Needs lock around `le.Ready`?

File: lease/lessor.go [hunk 2]
```diff
@@ -329,66 +332,82 @@ func (le *lessor) unsafeLeases() []*Lease {
...
+	go func() {
+		le.mu.RLock()
+		le.demotec = make(chan struct{})
```
Comment (gyuho) [inline lease/lessor.go]: `le.demotec` needs be protected by write lock.

**Comment by gyuho** [comment]
Have you benchmarked with real work workloads?

**Comment by mgates** [comment]
Still busy, but I blocked out some time at the end of next week to get it done. Sorry for the delay.

**Comment by mgates** [comment]
Ok - so I'm not sure what the best way to demonstarate this in a real-worldish way is, but here's what I did:

1) set up the test cluster with `goreman start`
2) put 1.9 million leases into etcd
3) Ran this snippet which tries to make a new lease every half second, and kills the leader after 10 seconds
```
while true; do ETCDCTL_API=3 etcdctl --command-timeout 0.2s --endpoints "http://127.0.0.1:2379,http://127.0.0.1:22379,http://127.0.0.1:32379" lease grant  60  > /dev/null ; echo `date` $? ; sleep 0.5; done & sleep 10; pkill -f  -- "[-]-listen-peer-urls $(ETCDCTL_API=2 etcdctl  member list  | grep isLeader=true | cut -f 3 -d "="| cut -f 1 -d " ")"
```

On the branch it had 1 timeout exceed error, but on master it 60, over the 61 seconds it took before it recovered.

Sorry it took so long - if these changes looks plausible to you, we can try to test it against our live, large workloads.

**Comment by mgates** [comment]
Hey folks - just wanted to check in about this - does it seem ok, at least in theory? If so, happy to work more on it.

File: lease/lessor.go [hunk 3]
```diff
@@ -329,66 +332,84 @@ func (le *lessor) unsafeLeases() []*Lease {
 	for _, l := range le.leaseMap {
 		leases = append(leases, l)
 	}
-	sort.Sort(leasesByExpiry(leases))
```
Comment (xiang90) [inline lease/lessor.go]: this change can be moved to a separate PR?

**Comment by xiang90** [comment]
@mgates 

Can you:

1. describe what problem (in which case what can happen and how this PR can help ) this PR solves
2. move the sorting change into another PR

I can understand this PR as it is, but I guess it is easier for other people to review if you do what I suggested.

**Comment by mgates** [comment]
Sure.
We had a problem where when we had a lot of leases (100kish), and a leader election happened the lessor was locked for a long time causing timeouts when we tried to grant a lease. This changes it so that the lessor is locked in batches, which allows for creation of leases while the leader is initializing the lessor. It still won't start expiring leases until it's all done rewriting them, though.

**Comment by xiang90** [comment]
> We had a problem where when we had a lot of leases (100kish), and a leader election happened the lessor was locked for a long time causing timeouts when we tried to grant a lease. This changes it so that the lessor is locked in batches, which allows for creation of leases while the leader is initializing the lessor. It still won't start expiring leases until it's all done rewriting them, though.

can you move this to the PR description and also put this into the commit message?

Thanks.

**Comment by mgates** [comment]
@xiang90 done, thanks a lot.

File: lease/lessor.go [hunk 4]
```diff
@@ -144,6 +144,9 @@ type lessor struct {
 	stopC chan struct{}
 	// doneC is a channel whose closure indicates that the lessor is stopped.
 	doneC chan struct{}
+
+	// when the lease pile-up reduction is done this is true
```
Comment (xiang90) [inline lease/lessor.go]: mention when this will be set to false?

**Comment by xiang90** [comment]
@mgates What happens when a lease is revoked while the go routine created in the promote is still running? is there a risk that the copy of the lease will be re-add into the heap?

**Comment by mgates** [comment]
That's a good question - I'm pretty sure that it would get updated, but since the revoke method deletes it from the lease map and deletes the appropriate keys, that wouldn't actually be a problem, and the lease would get cleaned up eventually.That's a good question - I'm pretty sure that it would get updated, but since the revoke method deletes it from the lease map and deletes the appropriate keys, that wouldn't actually be a problem, and the lease would get cleaned up eventually.

We'll write a test to confirm, though

**Comment by mgates** [comment]
Hey there, we looked into this more, and it doesn't seem possible to write a automated test of this behavior, we did some manual work and added some sleeps and feel good about the behavior - the lease TTL will get refreshed and possibly rewritten for the pile up avoidance, it'll get added to the heap, and will be ignored once the expiry checker gets to it, because it isn't in the lease map.

File: lease/lessor.go [hunk 5]
```diff
@@ -348,54 +353,70 @@ func (le *lessor) Leases() []*Lease {
...
+			return
+		}
 
-	// refresh the expiries of all leases.
-	for _, l := range le.leaseMap {
-		l.refresh(extend)
-		item := &LeaseWithTime{id: l.ID, expiration: l.expiry.UnixNano()}
-		heap.Push(&le.leaseHeap, item)
-	}
+		// adjust expiries in case of overlap
```
Comment (cyc115) [inline lease/lessor.go]: nit: white line

**Comment by wenjiaswe** [comment]
cc @jingyih
